File: Solutions/382-Linked List Random Node/382-Linked List Random Node.js
//Third fastest
var Solution = function(head) {
       this.head = head
};
Solution.prototype.getRandom = function() {
       let i = 2;
       let cur = this.head.next;
       let val = this.head.val;
       while(cur) {
              const ratio = 1/i
              if(Math.random() <= ratio ) val = cur.val;
              i++;
              cur = cur.next;
       }
       return val;
};

//Another
class Solution {
    constructor(head) {
        this.head = head;
    }
    
    getRandom() {
        let current = this.head;
        let nodeCount = 0;
        let randomVal = null;
        
        while (current !== null) {
            nodeCount++;
            
            if (Math.floor(Math.random() * nodeCount) === 0) {
                randomVal = current.val;
            }
            
            current = current.next;
        }
        
        return randomVal;
    }
}


//Second fastest
var Solution = function(head) {
    this.root = head;
    var curr = head, count = 0;
    while (curr) {
      curr = curr.next;
      count++;
    }
    this.count = count;
  };
  
  /**
   * Returns a random node's value.
   * @return {number}
   */
  Solution.prototype.getRandom = function() {
    var rand = Math.floor(Math.random() * this.count);
    var res = this.root;
    while (rand) {
      res = res.next;
      rand--;
    }
    return res.val;
  };

//Fastest
var Solution = function(head) {
    this.head = head;
    this.len = 1;
    while (head.next) {
        this.len++;
        head = head.next;
    }
    head.next = this.head;
};

/**
 * Returns a random node's value.
 * @return {number}
 */
Solution.prototype.getRandom = function() {
    let i = (Math.random() * this.len) >> 0;
    while(i--)  this.head = this.head.next;
    return this.head.val;
};

File: Solutions/382-Linked List Random Node/382-Linked List Random Node.rs
//Fastest
use rand::prelude::*;

struct Solution {
    len: usize,
    list: Option<Box<ListNode>>,
}

impl Solution {
    fn new(head: Option<Box<ListNode>>) -> Self {
        let mut cur = head.as_ref();
        let mut len = 0;
        while let Some(node) = cur {
            cur = node.next.as_ref();
            len += 1;
        }
        Self { len, list: head }
    }

    /** Returns a random node's value. */
    fn get_random(&self) -> i32 {
        if self.len == 0 {
            return -1;
        }
        if self.len == 1 {
            return self.list.as_ref().unwrap().val;
        }
        let mut rng = rand::thread_rng();
        let target = rng.gen_range(0, self.len);
        let mut cur = 0;
        let mut node = self.list.as_ref().unwrap();
        while cur < target {
            node = node.next.as_ref().unwrap();
            cur += 1;
        }
        node.val
    }
}

//Another
use rand::{thread_rng, Rng};
pub struct Solution {
    values: Vec<i32>,
}

impl Solution {
    /** @param head The linked list's head.
    Note that the head is guaranteed to be not null, so it contains at least one node. */
    fn new(head: Option<Box<ListNode>>) -> Self {
        let mut values = Vec::new();
        let mut node = head;
        while let Some(n) = node {
            values.push(n.val);
            node = n.next;
        }
        Self { values }
    }

    /** Returns a random node's value. */
    fn get_random(&self) -> i32 {
        let mut rng = thread_rng();
        self.values[rng.gen_range(0, self.values.len())]
    }
}